File: gate_pass/gatepass/report/gate_pass_report/gate_pass_report.py
# Copyright (c) 2025, sohail and contributors
# For license information, please see license.txt

# import frappe


# Copyright (c) 2025, Your Company
# For license information, please see license.txt

import frappe
from frappe.utils import flt

def execute(filters=None):
    if not filters:
        filters = {}

    columns = get_columns()
    data = get_data(filters)

    return columns, data


def get_columns():
    return [
        {"label": "Gate Pass No", "fieldname": "name", "fieldtype": "Link", "options": "Gate Pass", "width": 150},
        {"label": "Employee", "fieldname": "employee", "fieldtype": "Link", "options": "Employee", "width": 140},
        {"label": "Department", "fieldname": "department", "fieldtype": "Link", "options": "Department", "width": 140},
        {"label": "Pass Type", "fieldname": "pass_type", "fieldtype": "Data", "width": 100},
        {"label": "Floor", "fieldname": "floor", "fieldtype": "Data", "width": 120},
        {"label": "Status", "fieldname": "status", "fieldtype": "Data", "width": 100},
        {"label": "Issue Date", "fieldname": "issue_date", "fieldtype": "Datetime", "width": 160},
        {"label": "Return Date", "fieldname": "return_date", "fieldtype": "Datetime", "width": 160},
        {"label": "Item", "fieldname": "item", "fieldtype": "Link", "options": "Item", "width": 140},
        {"label": "Quantity", "fieldname": "qty", "fieldtype": "Float", "width": 100},
        {"label": "UOM", "fieldname": "uom", "fieldtype": "Link", "options": "UOM", "width": 100},
    ]


def get_data(filters):
    conditions = ""

    if filters.get("employee"):
        conditions += f" and gp.employee = '{filters.get('employee')}'"
    if filters.get("department"):
        conditions += f" and gp.department = '{filters.get('department')}'"
    if filters.get("from_date"):
        conditions += f" and gp.issue_date >= '{filters.get('from_date')}'"
    if filters.get("to_date"):
        conditions += f" and gp.issue_date <= '{filters.get('to_date')}'"

    data = frappe.db.sql(f"""
        SELECT
            gp.name,
            gp.employee,
            gp.department,
            gp.pass_type,
            gp.floor,
            gp.status,
            gp.issue_date,
            gp.return_date,
            child.item,
            child.qty,
            child.uom
        FROM
            `tabGate Pass` gp
        LEFT JOIN
            `tabGate Pass Item` child ON gp.name = child.parent
        WHERE
            gp.docstatus = 1 {conditions}
        ORDER BY
            gp.issue_date DESC
    """, as_dict=True)

    return data

File: gate_pass/gatepass/report/gate_pass_report/gate_pass_report.js
// Copyright (c) 2025, Sohail
// For license information, please see license.txt

/**
 * Gate Pass Report
 * Update #4: Added Gate Pass ID filter + filter reordering + UI polish
 */

frappe.query_reports["Gate Pass Report"] = {
    filters: [
        {
            fieldname: "gate_pass_id",
            label: __("Gate Pass ID"),
            fieldtype: "Link",
            options: "Gate Pass",
            width: 160
        },
        {
            fieldname: "employee",
            label: __("Employee ID"),
            fieldtype: "Link",
            options: "Employee",
            width: 160
        },
        {
            fieldname: "employee_name",
            label: __("Employee Name Contains"),
            fieldtype: "Data",
            width: 170
        },
        {
            fieldname: "department",
            label: __("Department"),
            fieldtype: "Link",
            options: "Department",
            width: 150
        },
        {
            fieldname: "purpose",
            label: __("Purpose Contains"),
            fieldtype: "Data",
            width: 170
        },
        {
            fieldname: "vehicle_no",
            label: __("Vehicle No"),
            fieldtype: "Data",
            width: 140
        },
        {
            fieldname: "status",
            label: __("Status"),
            fieldtype: "Select",
            options: ["", "Approved", "Rejected", "Pending"].join("\n"),
            width: 130
        },
        {
            fieldname: "gate_pass_type",
            label: __("Gate Pass Type"),
            fieldtype: "Select",
            options: ["", "In", "Out"].join("\n"),
            width: 130
        },
        {
            fieldname: "created_by",
            label: __("Created By"),
            fieldtype: "Link",
            options: "User",
            width: 150
        },
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.add_months(frappe.datetime.get_today(), -1)
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.get_today()
        }
    ]
};
